iroh-net/src/hp/derp/http/client.rs

```diff
@@ -832,6 +861,149 @@ impl Client {
         }
         Ok(())
     }
+
+    /// The tailscale code has a complicated way of dealing with the initial logging of the
+    /// `peers_present` map. Instead, we have an initial sleep of 5 seconds, and then print whatever
+    /// peers we know are present in that derp server. After that, we print the peers present every
+    /// 10 second interval.
+    ///
+    pub async fn run_mesh_client(
+        self,
+        ignore_key: key::node::PublicKey,
+        packet_forwarder_handler: PacketForwarderHandler<Client>,
+        cancel: CancellationToken,
+    ) -> anyhow::Result<()> {
+        let cancel_logging = cancel.clone();
+        tokio::select! {
+            _ = cancel.cancelled() => {
+                tracing::debug!("run_mesh_client loop cancelled");
+                Ok(())
+            },
+            res = self.run_mesh_client_0(ignore_key, packet_forwarder_handler, cancel_logging) => {
+                res
+            }
+        }
+    }
+
+    pub async fn run_mesh_client_0(
+        self,
+        ignore_key: key::node::PublicKey,
+        packet_forwarder_handler: PacketForwarderHandler<Client>,
+        cancel_logging: CancellationToken,
+    ) -> anyhow::Result<()> {
+        let redial_delay = Duration::from_secs(5);
+        let peers_present = PeersPresent::new();
+
+        let mut last_conn_gen = 0;
+
+        // connect to the remote server & request to watching the remote's state changes
+        loop {
+            let server_public_key = match self.watch_connection_changes().await {
+                Ok(key) => key,
+                Err(e) => {
+                    peers_present.clear().await;
+                    tracing::warn!("error connecting to derp server {e}");
+                    tokio::time::sleep(redial_delay).await;
+                    continue;
+                }
+            };
+
+            if server_public_key == ignore_key {
+                bail!("detected self-connect; closing this client");
+            }
+
+            tracing::info!("Connected to mesh derp server {server_public_key:?}");
+
+            let peers_logger = peers_present.clone();
+            let cancel = cancel_logging.clone();
+            tokio::spawn(async move {
+                let start = Instant::now() + PEERS_PRESENT_LOGGING_DELAY;
+                let mut status_logging_interval =
+                    tokio::time::interval_at(start, PEERS_PRESENT_LOGGING_INTERVAL);
+
+                loop {
+                    tokio::select! {
+                        biased;
+                        _ = cancel.cancelled() => {
+                            tracing::debug!("mesh update logging loop cancelled");
+                            break;
+                        },
+                        _ = status_logging_interval.tick() => {
+                        peers_logger.log_info().await;
```

Comment:
indent

Reply:
okay so there was something buggy with the way rust analyzer was interacting with vim and its plug ins, which is why rustfmt wasn't running... an hour and a half later i figured it out boooooooooo.